docs/SIL.rst

```diff
@@ -1538,20 +1538,23 @@ Class TBAA
 
 Class instances and other *heap object references* are pointers at the
 implementation level, but unlike SIL addresses, they are first class values and
-can be ``capture``-d and alias. Swift, however, is memory-safe and statically
+can be ``capture``-d and aliased. Swift, however, is memory-safe and statically
 typed, so aliasing of classes is constrained by the type system as follows:
 
 * A ``Builtin.NativeObject`` may alias any native Swift heap object,
   including a Swift class instance, a box allocated by ``alloc_box``,
   or a thick function's closure context.
   It may not alias natively Objective-C class instances.
-* A ``Builtin.UnknownObject`` may alias any class instance, whether Swift or
-  Objective-C, but may not alias non-class-instance heap objects.
-* Two values of the same class type ``$C`` may alias. Two values of related
-  class type ``$B`` and ``$D``, where there is a subclass relationship between
-  ``$B`` and ``$D``, may alias. Two values of unrelated class types may not
-  alias. This includes different instantiations of a generic class type, such
-  as ``$C<Int>`` and ``$C<Float>``, which currently may never alias.
+* A ``Builtin.UnknownObject`` or ``Builtin.BridgeObject`` may alias
+  any class instance, whether Swift or Objective-C, but may not alias
+  non-class-instance heap objects.
+* Two values of the same class type ``$C`` may alias. Two values of
+  related class type ``$B`` and ``$D``, where there is a subclass
+  relationship between ``$B`` and ``$D``, may alias. Two values of of
+  unrelated class types may not alias if typed access on both values
+  (via a ref_element_addr) are visible to the SIL program. This
+  includes different instantiations of a generic class type, such as
```

Comment:
> On Apr 11, 2016, at 10:36 PM, Mark Lacey <notifications@github.com> wrote:
> 
> In docs/SIL.rst <https://github.com/apple/swift/pull/2147#discussion_r59323025>:
> 
> > -* A ``Builtin.UnknownObject`` may alias any class instance, whether Swift or
> > -  Objective-C, but may not alias non-class-instance heap objects.
> > -* Two values of the same class type ``$C`` may alias. Two values of related
> > -  class type ``$B`` and ``$D``, where there is a subclass relationship between
> > -  ``$B`` and ``$D``, may alias. Two values of unrelated class types may not
> > -  alias. This includes different instantiations of a generic class type, such
> > -  as ``$C<Int>`` and ``$C<Float>``, which currently may never alias.
> > +* A ``Builtin.UnknownObject`` or ``Builtin.BridgeObject`` may alias
> > +  any class instance, whether Swift or Objective-C, but may not alias
> > +  non-class-instance heap objects.
> > +* Two values of the same class type ``$C`` may alias. Two values of
> > +  related class type ``$B`` and ``$D``, where there is a subclass
> > +  relationship between ``$B`` and ``$D``, may alias. Two values of of
> > +  unrelated class types may not alias if typed access on both values
> > +  (via a ref_element_addr) are visible to the SIL program. This
> > +  includes different instantiations of a generic class type, such as
> I'm not sure I know what you're getting at here. Can you clarify?
> 
> I might be reading it wrong, but as worded it sounds like I need to see the typed access for these to not alias, whereas if I don't see that (e.g. they are parameters), they do alias (perhaps because of casts through Builtin.{Native,Unknown,Bridge}Object.
> 

I came across situations where the stdlib casts to class types that aren’t statically related. I wanted to make sure those cases are covered by the spec, because I don’t think we want to say that the stdlib has undefined behavior.

I removed the line above from the latest rewrite because I think it’s was conflating the two kinds of TBAA. I think the issue is better covered by the side note below (basically, the optimizer can pass around invalid class types as long as it doesn’t dereference them):

A violation of the above aliasing rules only results in undefined
behavior if the aliased types are actually accessed *statically*
within Swift code. For example,
``_SwiftNativeNS[Array|Dictionary|String]`` classes alias with
``NS[Array|Dictionary|String]`` classes even though they are not
statically related. Since Swift never directly accesses stored
properties on the Foundation classes, this aliasing does not pose a
danger.

Andy